pkg/metrics/process/process_retriever_cached_darwin.go

```diff
@@ -0,0 +1,383 @@
+func (s *ProcessRetrieverCached) processesFromCache() (map[int32]psItem, error) {
+	s.cache.Lock()
+	defer s.cache.Unlock()
+
+	if s.cache.expired() {
+		psBin, err := exec.LookPath("ps")
+		if err != nil {
+			return nil, err
+		}
+		// it's easier to get the thread num per process from different call
+		processesThreads, err := s.getProcessThreads(psBin)
+		if err != nil {
+			return nil, err
+		}
+		// it's easier to get the thread num per process from different call
+		fullCmd, err := s.getProcessFullCmd(psBin)
+		if err != nil {
+			return nil, err
+		}
+		//get all processes and inject numThreads
+		items, err := s.retrieveProcesses(psBin)
+		if err != nil {
+			return nil, err
+		}
+		items = addThreadsAndCmdToPsItems(items, processesThreads, fullCmd)
+		s.cache.update(items)
+	}
+
+	return s.cache.items, nil
+}
+
+func addThreadsAndCmdToPsItems(items map[int32]psItem, processesThreads map[int32]int32, processesCmd map[int32]string) map[int32]psItem {
+	for pid, item := range items {
+		if numThreads, ok := processesThreads[pid]; ok {
+			item.numThreads = numThreads
+		}
+		if cmd, ok := processesCmd[pid]; ok {
+			item.cmdLine = cmd
+		}
+		items[pid] = item
+	}
+	return items
+}
+
+func (s *ProcessRetrieverCached) retrieveProcesses(psBin string) (map[int32]psItem, error) {
+
+	// get all processes info
+	args := []string{"ax", "-c", "-o", "pid,ppid,user,state,utime,stime,etime,rss,vsize,pagein,command"}
+	out, err := commandRunner(psBin, "", args...)
+	if err != nil {
+		return nil, err
+	}
+
+	lines := strings.Split(out, "\n")
+	items := make(map[int32]psItem)
+	for _, line := range lines[1:] {
+		var lineItems []string
+		for _, lineItem := range strings.Split(line, " ") {
+			if lineItem == "" {
+				continue
+			}
+			lineItems = append(lineItems, strings.TrimSpace(lineItem))
+		}
+		if len(lineItems) > 10 {
+			pid, _ := strconv.Atoi(lineItems[0])
+			ppid, _ := strconv.Atoi(lineItems[1])
+			user := lineItems[2]
+			state := lineItems[3]
+			utime := lineItems[4]
+			stime := lineItems[5]
+			etime := lineItems[6]
+			rss, _ := strconv.ParseInt(lineItems[7], 10, 64)
+			vsize, _ := strconv.ParseInt(lineItems[8], 10, 64)
+			pagein, _ := strconv.ParseInt(lineItems[9], 10, 64)
+			command := strings.Join(lineItems[10:], " ")
+
+			item := psItem{
+				pid:      int32(pid),
+				ppid:     int32(ppid),
+				username: user,
+				state:    state,
+				utime:    utime,
+				stime:    stime,
+				etime:    etime,
+				rss:      rss,
+				vsize:    vsize,
+				pagein:   pagein,
+				command:  command,
+			}
+			items[int32(pid)] = item
+		}
+	}
+	return items, nil
+}
+
+func (s *ProcessRetrieverCached) getProcessThreads(psBin string) (map[int32]int32, error) {
+	// get all processes info with threads
+	args := []string{"ax", "-M", "-c"}
+	out, err := commandRunner(psBin, "", args...)
+	if err != nil {
+		return nil, err
+	}
+
+	lines := strings.Split(out, "\n")
+	processThreads := make(map[int32]int32)
+	for _, line := range lines[1:] {
```

Comment:
could this panic?

Reply:
just verified and it will not https://play.golang.org/p/3sQCUZk8fA-  :)